src/items/recipes.zig

```
@@ -0,0 +1,226 @@
+		id = id[index + 1 ..];
+	}
+	const pattern = try parsePattern(allocator, id);
+	return .{
+		.amount = amount,
+		.pattern = pattern,
+	};
+}
+
+fn matchWithKeys(allocator: NeverFailingAllocator, target: []const u8, pattern: []const Segment, keys: *const std.StringHashMap([]const u8)) ![]std.StringHashMap([]const u8) {
+	var idx: usize = 0;
+	var newKeys = keys.clone() catch unreachable;
+	errdefer newKeys.deinit();
+
+	for(0.., pattern) |i, segment| {
+		switch(segment) {
+			.literal => |literal| {
+				if(literal.len + idx > target.len or !std.mem.eql(u8, target[idx .. idx + literal.len], literal)) {
+					return error.NoMatch;
+				}
+				idx += literal.len;
+			},
+			.symbol => |symbol| {
+				var indexes: main.List(usize) = .init(allocator);
+				defer indexes.deinit();
+				if(i + 1 < pattern.len) {
+					const nextSegment = pattern[i + 1];
+					var nextIndex = idx;
+					while(std.mem.indexOfPos(u8, target, nextIndex, nextSegment.literal)) |endIndex| {
+						indexes.append(endIndex);
+						nextIndex = endIndex + 1;
+					}
+				} else {
+					indexes.append(target.len);
+				}
+				if(indexes.items.len == 0) {
+					return error.NoMatch;
+				}
+				if(indexes.items.len == 1) {
+					if(newKeys.get(symbol)) |value| {
+						if(!std.mem.eql(u8, target[idx..indexes.items[0]], value)) {
+							return error.NoMatch;
+						}
+					} else {
+						newKeys.put(allocator.dupe(u8, symbol), allocator.dupe(u8, target[idx..indexes.items[0]])) catch unreachable;
+					}
+					idx = indexes.items[0];
+				} else {
+					defer newKeys.deinit();
+					var newKeyPairs: main.List(std.StringHashMap([]const u8)) = .init(allocator);
+					defer newKeyPairs.deinit();
+					for(indexes.items) |endIndex| {
+						if(matchWithKeys(allocator, target[endIndex..], pattern[i + 1 ..], &newKeys)) |newKeyMatches| {
+							newKeyPairs.appendSlice(newKeyMatches);
+							allocator.free(newKeyMatches);
+						} else |_| {}
```

Please don't ignore errors.

Again proper testing would have helped you find problem with this.